### PR TITLE
Fixed #27026 -- Fixed state initialization of bulk_create() objects if can_return_ids_from_bulk_insert.

### PR DESCRIPTION
**Pull request by sjoerdjob**
The `bulk_create` now also updates the `_state` field of a model
instance. This makes it possible to use it in `ManyToMany` relations.

### CODE REVIEW

File: tests/bulk_create/models.py [hunk 1]
```diff
@@ -6,6 +6,11 @@ class Country(models.Model):
     iso_two_letter = models.CharField(max_length=2)
 
 
+class Union(models.Model):
```
Comment (sjoerdjob) [inline tests/bulk_create/models.py]: This was the only thing I could think of that has a logical `ManyToMany` relation with `Countries`.

Come to think of it, another option might have been `Languages`, but that seems like a bit of a stretch.

If you have a better idea for a `ManyToMany` model for this test, please let me know.

File: django/db/models/query.py [hunk 2]
```diff
@@ -454,6 +454,8 @@ def bulk_create(self, objs, batch_size=None):
                         assert len(ids) == len(objs_without_pk)
                     for i in range(len(ids)):
                         objs_without_pk[i].pk = ids[i]
+                        objs_without_pk[i]._state.adding = False
+                        objs_without_pk[i]._state.db = self.db
```
Comment (charettes) [inline django/db/models/query.py]: It might be worth storing `objs_without_pk[i]` in an intermediary variable to avoid 3 key access.

Using `zip` would look be even better IMHO:

```python
db = self.db
for obj_without_pk, pk in zip(objs_without_pk, ids):
    obj_without_pk.pk = pk
    obj_without_pk._state.adding = False
    obj_without_pk._state.db = db
```

Comment (sjoerdjob) [inline django/db/models/query.py]: That reads so much more Pythonic, agreed. The caching of `self.db` seems overkill, but the rest makes a lot of sense.

File: django/db/models/query.py [hunk 3]
```diff
@@ -435,11 +435,12 @@ def bulk_create(self, objs, batch_size=None):
         if not objs:
             return objs
         self._for_write = True
-        connection = connections[self.db]
+        db = self.db
```
Comment (timgraham) [inline django/db/models/query.py]: Is this just a stylistic thing? If so, I don't see much advantage to it.

File: django/db/models/query.py [hunk 4]
```diff
@@ -452,8 +453,10 @@ def bulk_create(self, objs, batch_size=None):
                     ids = self._batched_insert(objs_without_pk, fields, batch_size)
                     if connection.features.can_return_ids_from_bulk_insert:
                         assert len(ids) == len(objs_without_pk)
-                    for i in range(len(ids)):
-                        objs_without_pk[i].pk = ids[i]
+                    for obj_without_pk, pk in zip(objs_without_pk, ids):
```
Comment (timgraham) [inline django/db/models/query.py]: The `zip(objs_without_pk, ids)` change seems unrelated to this fix? If it's more efficient or something, it would be better to make the change in a separate commit.

File: django/db/models/query.py (hunk 3 above)
Comment (charettes) [inline django/db/models/query.py]: @timgraham I was the one who suggested storing `db` in a locale variable to prevent multiple attribute lookups in the primary assignment loop. Might be overkill as you both pointed out.

File: django/db/models/query.py (hunk 4 above)
Comment (charettes) [inline django/db/models/query.py]: I suggested using zip now that we do multiple index lookips in `objs_without_pk` which we can assume to be relatively large given `bulk_create` was used in favor of `create`.

File: django/db/models/query.py (hunk 3 above)
Comment (timgraham) [inline django/db/models/query.py]: Oh, not sure. I'm not a micro-optimizer. :-)

Comment (charettes) [inline django/db/models/query.py]: @sjoerdjob let's remove the `db` optimization to keep the patch concise. On the other hand I still think we should use `zip`.

File: tests/bulk_create/models.py (hunk 1 above)
Comment (charettes) [inline tests/bulk_create/models.py]: Ideally we would the `m2m` field to one of the existing models of this test application to prevent the creation of an extra table on test database initialization. What about adding a self referential `ManyToMany` to `Country` named `neighboring`?

File: tests/bulk_create/tests.py [hunk 5]
```diff
@@ -217,3 +218,11 @@ def test_set_pk_and_query_efficiency(self):
...
         self.assertEqual(Country.objects.get(pk=countries[2].pk), countries[2])
         self.assertEqual(Country.objects.get(pk=countries[3].pk), countries[3])
+
+    @skipUnlessDBFeature('can_return_ids_from_bulk_insert')
+    def test_set_pk_and_allow_creating_related(self):
+        union_eu = Union.objects.create(name='EU')
+        country_nl = Country(name='Netherlands', iso_two_letter='NL')
+        with self.assertNumQueries(1):
+            Country.objects.bulk_create([country_nl])
+        union_eu.countries.add(country_nl)
```
Comment (charettes) [inline tests/bulk_create/tests.py]: As the underlying issue is state initialization and many-to-many assignment failure is only a symptom I suggest testing for `country_nl._state.adding` and `country_nl._state.db` instead. This should avoid defining the `Union` model at the same time.

**Comment by sjoerdjob** [comment]
I'll revert the `(self.)db` change tomorrow.

As for the `zip` vs `range/len`: my opinion would be that that change is sufficiently localized next to the 'fix' that it does not make to separate it into a commit on its own. But if you think it makes more sense, I'll oblige.

File: docs/releases/1.10.1.txt [hunk 6]
```diff
@@ -29,3 +29,6 @@ Bugfixes
 
 * Fixed the ``isnull`` lookup on a ``ForeignKey`` with its ``to_field``
   pointing to a ``CharField`` (:ticket:`26983`).
+
+* Fixed updating ``obj._state`` when doing ``bulk_create`` with a backend that
```
Comment (timgraham) [inline docs/releases/1.10.1.txt]: Made bulk_create() properly initialize model instances on backends such as PostgreSQL that support ...

File: docs/releases/1.10.1.txt [hunk 7]
```diff
@@ -29,3 +29,6 @@ Bugfixes
...
+  supports returning the IDs of the newly created records (:ticket:`27026`).
```
Comment (timgraham) [inline docs/releases/1.10.1.txt]: add: "so that many-to-many relationships can be used on the new objects."

File: django/db/models/query.py (hunk 3 above)
Comment (timgraham) [inline django/db/models/query.py]: sounds good to me

File: tests/bulk_create/tests.py [hunk 8]
```diff
@@ -217,3 +217,16 @@ def test_set_pk_and_query_efficiency(self):
         self.assertEqual(Country.objects.get(pk=countries[1].pk), countries[1])
         self.assertEqual(Country.objects.get(pk=countries[2].pk), countries[2])
         self.assertEqual(Country.objects.get(pk=countries[3].pk), countries[3])
+
+    @skipUnlessDBFeature('can_return_ids_from_bulk_insert')
```
Comment (sjoerdjob) [inline tests/bulk_create/tests.py]: Should we also add a testcase that if the database does not have the feature, the `.state` field is not updated?

File: docs/releases/1.10.1.txt [hunk 9]
```diff
@@ -29,3 +29,8 @@ Bugfixes
 
 * Fixed the ``isnull`` lookup on a ``ForeignKey`` with its ``to_field``
   pointing to a ``CharField`` (:ticket:`26983`).
+
+* Made `bulk_create()` properly initialize model instances on backends such as
+  PostgreSQL that support returning the IDs of the newly created records so
+  that many-to-many relationships can be used on the new objects
+  (:ticket:`27026`).
```
Comment (sjoerdjob) [inline docs/releases/1.10.1.txt]: Wondering what reads better:

"on backends that support returning the IDs of the newly created records -- such as PostgresSQL -- so that ..."

or what we have now.

**Comment by timgraham** [comment]
This looks sufficient. Merged in 3246d2b4bb981a8d782a349a99e9b89206028cee, thanks!
